File: miniprojects/battleship_project/board.js

class Board {
    constructor(n){
        this.size = n * n;
        this.grid = Array(n);
        for(let i = 0; i < this.grid.length; i++){
            this.grid[i] = Array(n).fill('N')
        }

    }
    size(){
        return this.size;
    }
    get(pos){
        return this.grid[pos[0]][pos[1]];
    }
    set(pos, value){
        this.grid[pos[0]][pos[1]] = value;
    }
    numShips(){
        let count = 0;
        for(let subArray of this.grid){
            for(let el of subArray){
                if(el === 'S') count++;
            }
        }
        return count;
    }
    attack(pos){
        const target = this.get(pos);
        if(target === 'S'){
            this.set(pos, 'H');
            console.log('You sunk my battleship!');
            return true;
        }else{
            this.set(pos, 'X');
            return false;
        }
    }
    placeRandomShips(){
        const totalShips = Math.ceil(this.size * 0.25);
        while(this.numShips() < totalShips){
            const row = Math.floor(Math.random() * this.grid.length);
            const col = Math.floor(Math.random() * this.grid.length);
            this.set([row, col], 'S');
        }
    }
    hiddenShipsGrid(){
        return this.deepCopy(this.grid);
    }
    deepCopy(grid){
        const newGrid = [];
        for(const ele of grid){
            if(Array.isArray(ele)){
                newGrid.push(this.deepCopy(ele));
            }else{
                if(ele === 'S') {
                    newGrid.push('N');
                }else{
                    newGrid.push(ele);
                }

            }
        }
        return newGrid;
    }
    static printGrid(grid){
        for(const subArray of grid){
            console.log(subArray.join('  '));
        }
    }
    print(){
        Board.printGrid(this.hiddenShipsGrid());
    }
    cheat(){
        Board.printGrid(this.grid);
    }

}

module.exports = Board;

File: AA_Exercise_Projects/W2D3/battleship_project/lib/board.rb
class Board
    attr_reader :size
  def initialize(n)
    @grid = Array.new(n) {Array.new(n,:N)}
    @size = n * n
  end
  def [](position) #position eg. [0,1]
    row = position[0]
    col = position[1]
    @grid[row][col]
  end
  def []=(position,val)
    row = position[0]
    col = position[1]
    @grid[row][col] = val
  end
  def num_ships
    count = 0
    @grid.each_with_index do |row,i|
        row.each_with_index do |col,j|
            count += 1 if @grid[i][j] == :S
        end
    end
    count
  end
  def attack(pos)
    if self[pos] == :S 
        self[pos] = :H
        puts 'you sunk my battleship!'
        return true
    else
        self[pos] = :X
        return false
    end
  end
  def place_random_ships
    n_ships = (@size * 25) / 100
    while self.num_ships < n_ships
        row = rand(0..@grid.length - 1)
        col = rand(0..@grid.length - 1)
        self[[row,col]] = :S
    end
  end
  def hidden_ships_grid
    @grid.map do |subarray|
        subarray.map do |ele|
            if ele == :S
                :N
            else
                ele
            end
        end
    end
  end
  def self.print_grid(grid)
    grid.each {|sub| puts sub.join(" ")}
  end
  def cheat
    Board.print_grid(@grid)
  end
  def print
    Board.print_grid(hidden_ships_grid)
  end
end
